File: valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  const stack = [];

  for (let i = 0; i < s.length; i++) {
    switch (s[i]) {
      case "(":
        stack.push(s[i]);
        break;
      case ")":
        if (stack.at(stack.length - 1) !== "(") {
          return false;
        }
        stack.pop();
        break;
      case "{":
        stack.push(s[i]);
        break;
      case "}":
        if (stack.at(stack.length - 1) !== "{") {
          return false;
        }
        stack.pop();
        break;
      case "[":
        stack.push(s[i]);
        break;
      case "]":
        if (stack.at(stack.length - 1) !== "[") {
          return false;
        }
        stack.pop();
        break;
      default:
        console.log(s[i]);
        break;
    }
  }
  if (stack.length) {
    return false;
  }
  return true;
};

console.log(isValid("["));

File: valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for i, char in enumerate(s):
            match char:
                case "(":
                    stack.append(char)
                case ")":
                    if i == 0:
                        return False
                    length = len(stack)
                    if length == 0:
                        return False
                    if stack[-1] != "(":
                        return False
                    stack.pop()
                case "{":
                    stack.append(char)
                case "}":
                    if i == 0:
                        return False
                    length = len(stack)
                    if length == 0:
                        return False
                    if stack[-1] != "{":
                        return False
                    stack.pop()
                case "[":
                    stack.append(char)
                case "]":
                    if i == 0:
                        return False
                    length = len(stack)
                    if length == 0:
                        return False
                    if stack[-1] != "[":
                        return False
                    stack.pop()
                case _:
                    pass

        if len(stack):
            return False

        return True


solution = Solution()
print(solution.isValid("(){}}{"))
